### PR TITLE
fix: connection could be called twice on set

### PR DESCRIPTION
**Pull request by hannojg**
### Details

This PR fixes cases where a call to `Onyx.set` could invoke a `Onyx.connect` listener twice. 
For `withOnyx` connections this isn't happening, as they are already guarded by previous equal comparisons.

The fix here is to add a map that will store all previous values that were send to the `Onyx.connect` callback. We can't reuse the cache here unfortunately, as it can contain values while no data has been send to the connection yet (and thus a comparison would lead to us not sending to the connection.

For useOnyx I had to add a special parameter to `Onyx.connect` called `alwaysNotify` which is false by default.
`useOnyx` might uses a selector to transform the data. The output of that selector can change, even though the input data stay the same.
Thus for `useOnyx` we always have to call the connect callback.

### Related Issues

https://github.com/Expensify/react-native-onyx/issues/569

### Automated Tests

- Added a test that explicitly tests for the edge case where a connection would be called twice
- Fixed previous test that now has a reduced connection callback count which is expected 

### Manual Tests

n/a

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
Note: pressing the send button might be buggy in the recordings, its a know issue [first reported here](https://expensify.slack.com/archives/C01GTK53T8Q/p1721039243892689) and is not related to this PR.

<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/a3ca2efc-78a4-4c7f-8b92-4f00b6fe7fa7

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/1d83dd10-3e12-41a2-b885-ad61120f2a92

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/07a76549-87ac-42b7-a65c-eb4f6148c9ce

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/3e962c75-ab38-44cb-b78b-a78adebcfca4

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/b031aa42-2241-4628-aa9f-d77032f10bf9

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/50ad6d61-170b-4b5a-af9f-414aa32de967

</details>

### CODE REVIEW

**Review by fabioh8010 — APPROVED** [review]

**Review by mountiny — DISMISSED** [review]
Changes look good to me too, waiting for a c+ testing

File: lib/types.ts [hunk 1]
```diff
@@ -263,6 +263,11 @@ type Collection<TKey extends CollectionKeyBase, TValue, TMap = never> = {
 /** Represents the base options used in `Onyx.connect()` method. */
 type BaseConnectOptions = {
     initWithStoredValues?: boolean;
+    /**
+     * If true, it will call the callback function even if the value might hasn't changed.
```
Comment (blazejkustra) [inline lib/types.ts:267]: I think you left an error here, right? "Value might hasn't changed"

**Review by blazejkustra — APPROVED** [review]

File: lib/types.ts (hunk 1 above)
Comment (hannojg) [reply lib/types.ts:267]: Hm maybe my wording wasn't precise here.
When this value is `true` it can happen that the Onyx.conenct callback gets called with the same value multiple times (basically bypass the improvement)

Do you have an idea how to rephrase it to make it more obvious / better understandable?

**Comment by hannojg** [comment]
hm, maybe one second before we continue, want to ask Fabio one thing who wrote the useOnxy hook. Maybe we don't need the `alwaysNotify` prop

File: lib/types.ts (hunk 1 above)
Comment (hannojg) [reply lib/types.ts:267]: Currently discussing if this param is really needed

https://expensify.slack.com/archives/C05LX9D6E07/p1721746491534399?thread_ts=1721742421.298419&cid=C05LX9D6E07

**Comment by hannojg** [comment]
Yes, thanks for being patient with me!
I figured that the `alwaysNotify` isn't needed.

There was this test that failed with the changes I made:

Screenshot 2024-07-23 at 17 46 23

The thing is that the test case was a bit broken. In the test we change the selector function reference. In a react application that will only happen when the hook gets called again aka. the component re-renders (if we follow all strict rules, which we do). 
The test was working earlier early because the Onyx.connect callback was called another time, which is what this PR fixes.

File: lib/OnyxUtils.ts [hunk 2]
```diff
@@ -72,6 +72,9 @@ let defaultKeyStates: Record<OnyxKey, OnyxValue<OnyxKey>> = {};
 let batchUpdatesPromise: Promise<void> | null = null;
 let batchUpdatesQueue: Array<() => void> = [];
 
+// Holds a map representing the latest data that was used to invoke a connection callback
```
Comment (marcaaron) [inline lib/OnyxUtils.ts:75]: **non blocking thought:** maybe better to explain why we need this and what it is used for.

File: lib/OnyxUtils.ts [hunk 3]
```diff
@@ -960,6 +969,15 @@ function sendDataToConnection<TKey extends OnyxKey>(mapping: Mapping<TKey>, valu
     // If we would pass undefined to setWithOnyxInstance instead, withOnyx would not set the value in the state.
     // withOnyx will internally replace null values with undefined and never pass null values to wrapped components.
     // For regular callbacks, we never want to pass null values, but always just undefined if a value is not set in cache or storage.
+    const valueToPass = value === null ? undefined : value;
```
Comment (marcaaron) [inline lib/OnyxUtils.ts:972]: **nitpick:** we can also update line 981 to use `valueToPass`

**Review by marcaaron — DISMISSED** [review]
Changes look good. No blockers from me.

File: tests/unit/useOnyxTest.ts [hunk 4]
```diff
@@ -276,6 +276,8 @@ describe('useOnyx', () => {
             expect(result.current[1].status).toEqual('loaded');
 
             selector = (entry: OnyxEntry<{id: string; name: string}>) => `id - ${entry?.id}, name - ${entry?.name} - selector changed`;
+            // In a react app we expect the selector ref to change during a rerender (see selectorRef/useLiveRef)
+            rerender(undefined);
 
             await act(async () => Onyx.merge(ONYXKEYS.TEST_KEY, {id: 'changed_id', name: 'changed_name'}));
```
Comment (fabioh8010) [inline tests/unit/useOnyxTest.ts:282]:
```suggestion
            await act(async () => waitForPromisesToResolve());

            expect(result.current[0]).toEqual('id - test_id, name - test_name - selector changed');
```

I think we can remove the Onyx.merge since the rerender() is enough to prove the selector changed

Comment (hannojg) [reply tests/unit/useOnyxTest.ts:282]: ah good idea, let me check!

Comment (hannojg) [reply tests/unit/useOnyxTest.ts]: We don't even need waitForPromisesToResolve() 😊 looks good?

**Review by fabioh8010 — APPROVED** [review]

**Comment by hannojg** [comment]
@mountiny I think we are good to merge!

**Review by marcaaron — APPROVED** [review]

**Comment by hannojg** [comment]
They are failing due to another bug in Onyx for which I have a PR to fix here:

- https://github.com/Expensify/react-native-onyx/pull/574
